src/main/scala/org/deepdive/ddlog/DeepDiveLogSemanticChecker.scala

```diff
@@ -34,28 +34,31 @@ object DeepDiveLogSemanticChecker extends DeepDiveLogHandler {
     checkFunctionDefined(stmt)
     checkVariableRelationSchema(stmt)
     checkNumberOfColumns(stmt)
+    checkOptionalModifier(stmt)
+  }
+
+  // iterate over all atoms contained in the body list and apply the checker
+  def checkBodyAtoms(checker: Atom => Unit)(bodies: List[Body]) : Unit = {
+    bodies foreach {
+      case a: Atom => checker(a)
+      case a: QuantifiedBody => checkBodyAtoms(checker)(a.bodies)
+      case _ =>
+    }
   }
 
   // check if relations in the body are defined
   def checkRelationDefined(stmt: Statement) {
     val stmtStr = DeepDiveLogPrettyPrinter.print(stmt)
-    def checkRelation(name: String) {
-      if (!(heads contains name))
-        error(stmt, s"""relation "${name}" is not defined""")
+    def checkRelation(a: Atom) {
+      if (!(heads contains a.name))
+        error(stmt, s"""relation "${a.name}" is not defined""")
     }
+    def check = checkBodyAtoms(checkRelation)(_)
```

Comment:
IMO, this syntax looks more complicated than it needs to be... what's the advantage of this one over just the simple function prototype?